core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -215,14 +227,110 @@ public boolean isAvailable() {
       return file.exists();
    }
 
+   private boolean isCorruptData() throws IOException {
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      long pos = filePos;
+      buf = readChannel(buf, pos, KEY_POS_LATEST);
+      if (buf.remaining() > 0)
+         return false;
+      buf.flip();
+      FileEntry fe = new FileEntry(pos, buf);
+      // Explicitly calculate the correct size as incorrect value stored in fe.size
+      int actualSize = KEY_POS_LATEST + fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen;
+
+      // Attempt to read the next entry in the file
+      pos += actualSize;
+      buf = readChannel(buf, pos, KEY_POS_LATEST);
+
+      // Only single FileEntry present with correct size, so data is not corrupt
+      if (buf.remaining() > 0)
+         return false;
+
+      buf.flip();
+      fe = new FileEntry(pos, buf);
+      channel.position(filePos);
+      return fe.size != fe.keyLen + fe.dataLen + fe.metadataLen + fe.internalMetadataLen + 28 + (fe.expiryTime != -1 ? TIMESTAMP_BYTES : 0);
+   }
+
+   private void migrateCorruptData() throws IOException {
+      String cacheName = ctx.getCache().getName();
+      PERSISTENCE.startRecoveringCorruptPersistenceData(cacheName);
+      File newFile = new File(file.getParentFile(), cacheName + "_new.dat");
+
+      long oldFilePos = MAGIC_LATEST.length;
+      ByteBuffer buf = ByteBuffer.allocate(KEY_POS_LATEST);
+      try (FileChannel newChannel = SecurityActions.openFileChannel(newFile)) {
+         //Write Magic
+         newChannel.truncate(0);
+         newChannel.write(ByteBuffer.wrap(MAGIC_LATEST), 0);
+
+         while (true) {
+            buf = readChannel(buf, oldFilePos, KEY_POS_LATEST);
+            if (buf.remaining() > 0)
+               break;
+
+            buf.flip();
+
+            FileEntry fe = new FileEntry(oldFilePos, buf);
+            oldFilePos += KEY_POS_LATEST;
+            // Explicitly calculate the correct size as incorrect value stored in fe.size
+            // Meta length incorrectly added 16 bytes, regardless of whether fe.expiryTime > 0
+            int metaLen = fe.metadataLen > 0 ? fe.metadataLen - TIMESTAMP_BYTES : 0;
+
+            // Read old entry content and then write
+            ByRef.Long offset = new ByRef.Long(oldFilePos);
+
+            buf = readChannelUpdateOffset(buf, offset, fe.keyLen);
+            org.infinispan.commons.io.ByteBuffer key = readIntoByteBuffer(buf);
```

Comment:
shouldn't the key, value, and metadata be unmarshaller to check if it is compatible with the current marshaller?

Reply:
We could do this, but it will obviously be slower. The migration logic in general assumes that the user hasn't changed the global marshaller configuration when migrating between versions. This code is only going to be hit if the user is using the default protostream marshaller as `migrateFromV11` had (another) bug that means it would always fail if another marshaller was configured. This was due to the old `loadLegacyObject` implementation always attempting to read from the user marshaller first and only catching `IllegalArgumentException`. Other marshallers will fail with a different exception, for example JavaSerializationMarshaller throws `StreamCorruptedException`.

Offending code:

```
   @SuppressWarnings("unchecked")
   private <T> T loadLegacyObject(ByteBuffer buf, ByRef.Long offset, int length) throws ClassNotFoundException, IOException {
      buf = readChannelUpdateOffset(buf, offset, length);
      // Read using raw user marshaller without MarshallUserObject wrapping
      byte[] bytes = buf.array();
      Marshaller marshaller = ctx.getPersistenceMarshaller().getUserMarshaller();
      try {
         return (T) marshaller.objectFromByteBuffer(bytes, 0, length);
      } catch (IllegalArgumentException e) {
         // For internal cache key/values and custom metadata we need to use the persistence marshaller
         return (T) ctx.getPersistenceMarshaller().objectFromByteBuffer(bytes, 0, length);
      }
   }
```

Reply:
safer is better IMO.
since the old file is deleted after the migration, it is better to know of any marshalling issues during the migration than after.